docs/functionaldatatypes/zset.md

```diff
@@ -0,0 +1,276 @@
+
+```scala mdoc
+val sum: ZSet[Int, Double] =
+  die.zipWith(die)(_ + _)
+```
+
+The resulting set has eleven elements, corresponding to the possible sums from two dice rolls of two to twelve. The measure for each value corresponds to the probability of that sum being rolled.
+
+So for example the probability of seven is about 16.7% whereas the probability of rolling two or twelve is only about 2.8%.
+
+We can take this even further using the `flatMap`.
+
+The `zipWith` operator lets us combine two sets when the sets are independent of each other. With the `flatMap` operator we can create a new set based on each element of the first set and then "flatten" the resulting set down.
+
+We could rewrite the example above using `flatMap` and a _for comprehension_ like this:
+
+```scala mdoc:nest
+val sum: ZSet[Int, Double] =
+  for {
+    a <- die
+    b <- die
+  } yield a + b
+```
+
+This allows us to express conditional logic, such as the probability distribution of outcomes for a subsequent event depending on the outcome of a previous event.
+
+## Transforming Elements
+
+We can transform set elements using the `map` operator.
+
+For example, we could add one to each value in a set like this:
+
+```scala mdoc
+val setPlusOne: ZSet[Int, Boolean] =
+  set.map(_ + 1)
+```
+
+This requires a `Commutative` instance defined for the `Sum` of the measure type so that we can combine the measures if we map two elements to the same value.
+
+## Transforming The Measure
+
+We can also transform the measure we use to represent how many times an item appears using the `transform` operator. For example, perhaps we have constructed a set where the measure is an `Int` but we would like to transform it to a `Double` so we can compose it with other `ZSet` values with `Double` as the measure type.
+
+We can do that with `transform` like this:
+
+```scala mdoc
+def constant(n: Int): ZSet[Int, Double] =
+  ZSet(n).transform(_.toDouble)
+```
+
+## Extracting ZSet Values
+
+If the number of times a value appears in the set is a natural number then we can use normal collection operators to fold over its values, treating it much like a normal set
+
+```scala mdoc
+val fruits: MultiSet[String] =
+  ZSet.fromIterable(List("apple", "banana", "orange", "apple", "orange"))
+
+val fruitCount: Int =
+  fruits.foldLeft(0)((n, _) => n + 1)
+```
+
+This will return `5` because there are five elements in the set we are folding over.
+
+We can convert a `ZSet` back to a `Map` from elements to a measure of how many times they appear using the `toMap` operator.
+
+We can also always convert a `ZSet` back to a `Set` using the `toSet` operator. Any elements in the `ZSet` with a measure equal to zero based on an `Identity` defined on the `Sum` of the measure type will be discarded from the resulting set.
```

Comment:
hopefully this works well for a Double measures and numbers almost equal to 0.0 but distinct from 0.0.

Reply:
Yeah maybe we should require that the measure type be `Boolean` to convert it back so the user has to provide a function. I do think if the user is using a data type that is subject to floating point precision like that they are probably aware of the issues with it.